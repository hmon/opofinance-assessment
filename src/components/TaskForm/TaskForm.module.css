.TaskForm {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 16px;
  width: 100%;
  align-self: flex-start;

  @media (min-width: 768px) {
    width: 25%;
  }
}
