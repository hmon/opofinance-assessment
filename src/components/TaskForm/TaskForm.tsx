import React, { useState } from "react";
import useTaskStore from "../../store/task.ts";
import { TaskStatus } from "../../types";
import { TextField } from "../TextField/TextField.tsx";
import { Button } from "../Button/Button.tsx";

import styles from "./TaskForm.module.css";

const TaskForm: React.FC = () => {
  const addTask = useTaskStore((state) => state.addTask);
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');
  const [dueDate, setDueDate] = useState('');

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (!title || !description || !dueDate) return;

    addTask({
      title,
      description,
      status: TaskStatus.Pending,
      dueDate,
    });

    setTitle('');
    setDescription('');
    setDueDate('');
  };

  return (
    <form className={styles.TaskForm} onSubmit={handleSubmit} data-testid="task-form">
      <TextField
        label="Title"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
      />

      <TextField
        label="Description"
        value={description}
        onChange={(e) => setDescription(e.target.value)}
      />

      <TextField
        label="Due Date"
        type="date"
        value={dueDate}
        onChange={(e) => setDueDate(e.target.value)}
      />

      <Button type="submit">Add Task</Button>
    </form>
  );
};

export default TaskForm;
