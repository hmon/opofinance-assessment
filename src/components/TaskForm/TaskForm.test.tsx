import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import TaskForm from './TaskForm';

describe('TaskForm Component', () => {
  test('renders TaskForm component', () => {
    render(<TaskForm />);
    const taskFormElement = screen.getByTestId('task-form');
    expect(taskFormElement).toBeInTheDocument();
    expect(taskFormElement).toMatchSnapshot();
  });

  test('calls addTask when form is submitted', () => {
    render(<TaskForm />);

    fireEvent.change(screen.getByLabelText(/title/i), { target: { value: 'New Task' } });
    fireEvent.change(screen.getByLabelText(/description/i), { target: { value: 'New Description' } });
    fireEvent.change(screen.getByLabelText(/due date/i), { target: { value: '2023-12-31' } });

    fireEvent.click(screen.getByRole('button', { name: /add task/i }));

    expect(screen.getByLabelText(/title/i)).toHaveValue('');
    expect(screen.getByLabelText(/description/i)).toHaveValue('');
    expect(screen.getByLabelText(/due date/i)).toHaveValue('');
  });

  test('does not call addTask if form is incomplete', () => {
    render(<TaskForm />);

    fireEvent.change(screen.getByLabelText(/title/i), { target: { value: 'New Task' } });
    fireEvent.change(screen.getByLabelText(/description/i), { target: { value: '' } });
    fireEvent.change(screen.getByLabelText(/due date/i), { target: { value: '2023-12-31' } });

    fireEvent.click(screen.getByRole('button', { name: /add task/i }));

    expect(screen.getByLabelText(/title/i)).toHaveValue('New Task');
    expect(screen.getByLabelText(/description/i)).toHaveValue('');
    expect(screen.getByLabelText(/due date/i)).toHaveValue('2023-12-31');
  });
});
