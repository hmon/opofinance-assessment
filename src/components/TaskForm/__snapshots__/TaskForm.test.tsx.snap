// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskForm Component > renders TaskForm component 1`] = `
<form
  class="_TaskForm_e8b683"
  data-testid="task-form"
>
  <div
    class="_TextField_da86b6"
  >
    <label
      for=":r0:"
    >
      Title
    </label>
    <input
      class="_input_da86b6"
      id=":r0:"
      value=""
    />
  </div>
  <div
    class="_TextField_da86b6"
  >
    <label
      for=":r1:"
    >
      Description
    </label>
    <input
      class="_input_da86b6"
      id=":r1:"
      value=""
    />
  </div>
  <div
    class="_TextField_da86b6"
  >
    <label
      for=":r2:"
    >
      Due Date
    </label>
    <input
      class="_input_da86b6"
      id=":r2:"
      type="date"
      value=""
    />
  </div>
  <button
    class="_Button_2a1efe"
    type="submit"
  >
    Add Task
  </button>
</form>
`;
